pystapi-client/src/pystapi_client/client.py

```diff
@@ -0,0 +1,333 @@
+from collections.abc import Callable, Iterable, Iterator
+from typing import (
+    Any,
+)
+
+from httpx import Request
+from httpx._types import TimeoutTypes
+from pydantic import AnyUrl
+from stapi_pydantic import Link, Order, OrderCollection, Product, ProductsCollection
+
+from pystapi_client.conformance import ConformanceClasses
+from pystapi_client.exceptions import APIError
+from pystapi_client.stapi_api_io import StapiIO
+from pystapi_client.warnings import NoConformsTo
+
+DEFAULT_LINKS = [
+    {
+        "endpoint": "/conformance",
+        "rel": "conformance",
+        "method": "GET",
+    },
+    {
+        "endpoint": "/products",
+        "rel": "products",
+        "method": "GET",
+    },
+]
+
+
+class Client:
+    """A Client for interacting with the root of a STAPI
+
+    Instances of the ``Client`` class provide a convenient way of interacting
+    with STAPI APIs that conform to the [STAPI API spec](https://github.com/stapi-spec/stapi-spec).
+    """
+
+    stapi_io: StapiIO
+    _conforms_to: list[str]
+    _links: list[Link]
+
+    def __repr__(self) -> str:
+        return f"<Client {self.stapi_io.root_url}>"
+
+    @classmethod
+    def open(
+        cls,
+        url: str,
+        headers: dict[str, str] | None = None,
+        parameters: dict[str, Any] | None = None,
+        request_modifier: Callable[[Request], Request] | None = None,
+        timeout: TimeoutTypes | None = None,
+    ) -> "Client":
+        """Opens a STAPI API client
+
+        Args:
+            url : The URL of a STAPI API.
+            headers : A dictionary of additional headers to use in all requests
+                made to any part of this STAPI API.
+            parameters: Optional dictionary of query string parameters to
+                include in all requests.
+            request_modifier: A callable that either modifies a `Request` instance or
+                returns a new one. This can be useful for injecting Authentication
+                headers and/or signing fully-formed requests (e.g. signing requests
+                using AWS SigV4).
+
+                The callable should expect a single argument, which will be an instance
+                of :class:`requests.Request`.
+
+                If the callable returns a `requests.Request`, that will be used.
+                Alternately, the callable may simply modify the provided request object
+                and return `None`.
+            timeout: Optional float or (float, float) tuple following the semantics
+              defined by `Requests
+              <https://requests.readthedocs.io/en/latest/api/#main-interface>`__.
+
+        Return:
+            client : A :class:`Client` instance for this STAPI API
+        """
+        client = Client()
+        client.stapi_io = StapiIO(
+            root_url=AnyUrl(url),
+            headers=headers,
+            parameters=parameters,
+            request_modifier=request_modifier,
+            timeout=timeout,
+        )
```

Comment:
Little thing, but _to me_ it makes more sense to pass in theses `stapi_io` to the initializer, to make it clear that they're instance methods:

```suggestion
        stapi_io = StapiIO(
            root_url=AnyUrl(url),
            headers=headers,
            parameters=parameters,
            request_modifier=request_modifier,
            timeout=timeout,
        )
        client = Client(stapi_io)
```